extendr-ffi/src/altrep.rs

```diff
@@ -0,0 +1,399 @@
+/// Must copy up to n elements, starting at 0-based index i, into `buf[n]`, and return the number of elements copied. The default method defers to `RAW_ELT` and hence the Elt method above.
+/// Versions: Appeared in 3.6.0.
+pub type R_altraw_Get_region_method_t = ::std::option::Option<
+    unsafe extern "C" fn(arg1: SEXP, arg2: R_xlen_t, arg3: R_xlen_t, arg4: *mut Rbyte) -> R_xlen_t,
+>;
+/// Must return `x[i]` for the 0-based index i. The default method returns `COMPLEX(x)[i]`.
+pub type R_altcomplex_Elt_method_t =
+    ::std::option::Option<unsafe extern "C" fn(arg1: SEXP, arg2: R_xlen_t) -> Rcomplex>;
+/// Must copy up to n elements, starting at 0-based index i, into `buf[n]`, and return the number of elements copied. The default method defers to COMPLEX_ELT and hence the Elt method above.
+/// Versions: Appeared in 3.6.0.
+pub type R_altcomplex_Get_region_method_t = ::std::option::Option<
+    unsafe extern "C" fn(
+        arg1: SEXP,
+        arg2: R_xlen_t,
+        arg3: R_xlen_t,
+        arg4: *mut Rcomplex,
+    ) -> R_xlen_t,
+>;
+/// Must return `x[i]` (for 0-based i) or signal an error. The default method always raises an error.
+pub type R_altstring_Elt_method_t =
+    ::std::option::Option<unsafe extern "C" fn(arg1: SEXP, arg2: R_xlen_t) -> SEXP>;
+/// Must set `x[i] <- v` (for 0-based i), even if it’s not compatible with the chosen compact representation, or signal an error. The default method always raises an error.
+pub type R_altstring_Set_elt_method_t =
+    ::std::option::Option<unsafe extern "C" fn(arg1: SEXP, arg2: R_xlen_t, arg3: SEXP)>;
+/// Must return one of the sortedness constants: SORTED_DECR, UNKNOWN_SORTEDNESS, SORTED_INCR. See STRING_IS_SORTED. The default method always returns UNKNOWN_SORTEDNESS.
+pub type R_altstring_Is_sorted_method_t =
+    ::std::option::Option<unsafe extern "C" fn(arg1: SEXP) -> ::std::os::raw::c_int>;
+/// Must return 1 if x is known not to store any missing values (see STRING_NO_NA), otherwise 0 (including if it’s unknown). The default method always returns 0.
+pub type R_altstring_No_NA_method_t =
+    ::std::option::Option<unsafe extern "C" fn(arg1: SEXP) -> ::std::os::raw::c_int>;
+/// Must return `x[[i]]` for the 0-based index i or raise an error.
+pub type R_altlist_Elt_method_t =
+    ::std::option::Option<unsafe extern "C" fn(arg1: SEXP, arg2: R_xlen_t) -> SEXP>;
+/// Must set `x[[i]] <- v` for the 0-based index i, altering x, or raise an error.
+pub type R_altlist_Set_elt_method_t =
+    ::std::option::Option<unsafe extern "C" fn(arg1: SEXP, arg2: R_xlen_t, arg3: SEXP)>;
+
+extern "C" {
+    /// Returns a nonzero value if x is a “compact representation” (ALTREP) value. Access to the data pointers of ALTREP values may be problematic and should be replaced with access to individual elements (see REAL_ELT) or vector subregions (see REAL_GET_REGION). See also: <https://svn.r-project.org/R/branches/ALTREP/ALTREP.html>
```

Comment:
Where does all these comments come from?

Reply:
https://aitap.codeberg.page/R-api/ 

I can remove or add links, preferences?

Reply:
I like docs. And theirs are good. Either way is fine.